File: Algorithms/Sorting/Bucket_Sort/BucketSort.cs
using System;
using System.Collections;
using System.Collections.Generic;

class BucketSort {

	public static void bucketSort(float []arr, int numOfBuckets)
	{
		if (numOfBuckets <= 0)
		return;

		// create empty buckets
		List<float>[] buckets = new List<float>[numOfBuckets];
    	for (int i = 0; i < numOfBuckets; i++) 
		{
		buckets[i] = new List<float>();
		}

		// distributing elements to different buckets
		for (int i = 0; i < numOfBuckets; i++) 
		{
		float temp = arr[i] * numOfBuckets;
		buckets[(int)temp].Add(arr[i]);
		}

		// Sorting is done here
		for (int i = 0; i < numOfBuckets; i++) 
		{
		buckets[i].Sort();
		}

		// Concatination is done here
		int index = 0;
		for (int i = 0; i < numOfBuckets; i++) 
		{
			for (int j = 0; j < buckets[i].Count; j++)
			{
				arr[index++] = buckets[i][j];
			}
		}
	}


	// Main method
	public static void Main()
	{
		float []arr = new float[] { 0.276F,  0.612F,0.119F, 0.7F,0.99F, 0.448F};

		int n = arr.Length;
		bucketSort(arr, n);

		Console.WriteLine("Sorted array is ");
		foreach(float element in arr) 
		{
			Console.Write(element + " ");
		}
	}
}

File: Algorithms/Sorting/Bucket_Sort/BucketSort.go
/*
Bucket sort, or bin sort, is a sorting algorithm that works by distributing the elements of an array into a number of buckets. Each bucket is then sorted individually, either using a different sorting algorithm, or by recursively applying the bucket sorting algorithm.
*/

package main

import (
	"fmt"
	"os"
	"strconv"
)

// Method for Insertion Sort containig array in Float
func insertionSort(array []float64) {
	for i := 0; i < len(array); i++ {
		temp := array[i]
		j := i - 1
		for ; j >= 0 && array[j] > temp; j-- {
			array[j+1] = array[j]
		}
		array[j+1] = temp
	}
}

// Method for bucketSort
func bucketSort(array []float64, bucketSize int) []float64 {
	var max, min float64
	for _, n := range array {
		if n < min {
			min = n
		}
		if n > max {
			max = n
		}
	}
	nBuckets := int(max-min)/bucketSize + 1
	buckets := make([][]float64, nBuckets)
	for i := 0; i < nBuckets; i++ {
		buckets[i] = make([]float64, 0)
	}
	// for loop to iterate the range of an array
	for _, n := range array {
		idx := int(n-min) / bucketSize
		buckets[idx] = append(buckets[idx], n)
	}

	sorted := make([]float64, 0)
	for _, bucket := range buckets {
		if len(bucket) > 0 {
			insertionSort(bucket)
			sorted = append(sorted, bucket...)
		}
	}
	// returns sorted array
	return sorted
}

// main method from where program starts execution
func main() {

	
	// static array
	array :=[]float64{4, 1, 1, 3, 2, 2, 7, -6, 0}
	for _, arg := range os.Args[1:] {
		if n, err := strconv.ParseFloat(arg, 64); err == nil {
			array = append(array, n)
		}
	}
	fmt.Printf("Before Bucket Sort %v\n", array)
	// calling method bucket sort
	array = bucketSort(array, 5)
	fmt.Printf("After Bucket Sort %v\n", array)
}
